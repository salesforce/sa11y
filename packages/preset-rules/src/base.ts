/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

import { getA11yConfig, RuleInfo } from './rules';

export const baseRulesInfo: RuleInfo = new Map(
    Object.entries({
        'area-alt': { priority: 'P1', wcagSC: '1.1.1', wcagLevel: 'A' },
        'aria-allowed-attr': { priority: 'P1', wcagSC: '4.1.2', wcagLevel: 'A' },
        'aria-allowed-role': { priority: 'P1', wcagSC: '4.1.2', wcagLevel: 'A' },
        'aria-command-name': { priority: 'P1', wcagSC: '4.1.2', wcagLevel: 'A' },
        'aria-dialog-name': { priority: 'P1', wcagSC: '4.1.2', wcagLevel: 'A' },
        'aria-hidden-body': { priority: 'P1', wcagSC: '4.1.2', wcagLevel: 'A' },
        'aria-hidden-focus': { priority: 'P1', wcagSC: '4.1.2', wcagLevel: 'A' },
        'aria-input-field-name': { priority: 'P1', wcagSC: '4.1.2', wcagLevel: 'A' },
        'aria-meter-name': { priority: 'P1', wcagSC: '4.1.2', wcagLevel: 'A' },
        'aria-progressbar-name': { priority: 'P1', wcagSC: '4.1.2', wcagLevel: 'A' },
        'aria-required-attr': { priority: 'P1', wcagSC: '4.1.2', wcagLevel: 'A' },
        'aria-required-children': { priority: 'P1', wcagSC: '4.1.2', wcagLevel: 'A' },
        'aria-required-parent': { priority: 'P1', wcagSC: '4.1.2', wcagLevel: 'A' },
        'aria-roledescription': { priority: 'P3', wcagSC: '4.1.2', wcagLevel: 'A' },
        'aria-roles': { priority: 'P1', wcagSC: '4.1.2', wcagLevel: 'A' },
        'aria-toggle-field-name': { priority: 'P1', wcagSC: '4.1.2', wcagLevel: 'A' },
        'aria-tooltip-name': { priority: 'P1', wcagSC: '4.1.2', wcagLevel: 'A' },
        'aria-treeitem-name': { priority: 'P1', wcagSC: '4.1.2', wcagLevel: 'A' },
        'aria-valid-attr': { priority: 'P3', wcagSC: '4.1.2', wcagLevel: 'A' },
        'aria-valid-attr-value': { priority: 'P3', wcagSC: '4.1.2', wcagLevel: 'A' },
        'audio-caption': { priority: 'P1', wcagSC: '1.2.2', wcagLevel: 'A' },
        'autocomplete-valid': { priority: 'P2', wcagSC: '1.3.5', wcagLevel: 'AA' },
        'avoid-inline-spacing': { priority: 'P3', wcagSC: '1.4.12', wcagLevel: 'AA' },
        'blink': { priority: 'P1', wcagSC: '2.2.2', wcagLevel: 'A' },
        'button-name': { priority: 'P1', wcagSC: '4.1.2', wcagLevel: 'A' },
        'bypass': { priority: 'P3', wcagSC: '2.4.1', wcagLevel: 'A' },
        'color-contrast': { priority: 'P1', wcagSC: '1.4.3', wcagLevel: 'AA' },
        'definition-list': { priority: 'P2', wcagSC: '1.3.1', wcagLevel: 'A' },
        'dlitem': { priority: 'P2', wcagSC: '1.3.1', wcagLevel: 'A' },
        'document-title': { priority: 'P2', wcagSC: '2.4.2', wcagLevel: 'A' },
        'duplicate-id': { priority: 'P3', wcagSC: '4.1.1', wcagLevel: 'A' },
        'duplicate-id-active': { priority: 'P3', wcagSC: '4.1.1', wcagLevel: 'A' },
        'duplicate-id-aria': { priority: 'P1', wcagSC: '4.1.2', wcagLevel: 'A' },
        'form-field-multiple-labels': { priority: 'P2', wcagSC: '2.5.3', wcagLevel: 'A' },
        'frame-focusable-content': { priority: 'P1', wcagSC: '2.1.1', wcagLevel: 'A' },
        'frame-title': { priority: 'P3', wcagSC: '2.4.2', wcagLevel: 'A' },
        'html-has-lang': { priority: 'P2', wcagSC: '3.1.1', wcagLevel: 'A' },
        'html-lang-valid': { priority: 'P2', wcagSC: '3.1.1', wcagLevel: 'A' },
        'html-xml-lang-mismatch': { priority: 'P2', wcagSC: '3.1.1', wcagLevel: 'A' },
        'image-alt': { priority: 'P1', wcagSC: '1.1.1', wcagLevel: 'A' },
        'input-button-name': { priority: 'P1', wcagSC: '4.1.2', wcagLevel: 'A' },
        'input-image-alt': { priority: 'P1', wcagSC: '1.1.1', wcagLevel: 'A' },
        'label': { priority: 'P1', wcagSC: '3.3.2', wcagLevel: 'A' },
        'label-title-only': { priority: 'P1', wcagSC: '3.3.2', wcagLevel: 'A' },
        'link-name': { priority: 'P1', wcagSC: '4.1.2', wcagLevel: 'A' },
        'list': { priority: 'P2', wcagSC: '1.3.1', wcagLevel: 'A' },
        'listitem': { priority: 'P2', wcagSC: '1.3.1', wcagLevel: 'A' },
        'marquee': { priority: 'P1', wcagSC: '2.2.2', wcagLevel: 'A' },
        'meta-refresh': { priority: 'P1', wcagSC: '2.2.1', wcagLevel: 'A' },
        'meta-viewport': { priority: 'P2', wcagSC: '1.4.4', wcagLevel: 'AA' },
        'nested-interactive': { priority: 'P1', wcagSC: '4.1.2', wcagLevel: 'A' },
        'object-alt': { priority: 'P1', wcagSC: '1.1.1', wcagLevel: 'A' },
        'presentation-role-conflict': { priority: 'P3', wcagSC: '4.1.1', wcagLevel: 'A' },
        'role-img-alt': { priority: 'P1', wcagSC: '1.1.1', wcagLevel: 'A' },
        'scope-attr-valid': { priority: 'P1', wcagSC: '1.3.1', wcagLevel: 'A' },
        'scrollable-region-focusable': { priority: 'P1', wcagSC: '2.1.1', wcagLevel: 'A' },
        'select-name': { priority: 'P1', wcagSC: '4.1.2', wcagLevel: 'A' },
        'server-side-image-map': { priority: 'P1', wcagSC: '2.1.1', wcagLevel: 'A' },
        'svg-img-alt': { priority: 'P1', wcagSC: '1.1.1', wcagLevel: 'A' },
        'tabindex': { priority: 'P2', wcagSC: '2.4.3', wcagLevel: 'A' },
        'td-headers-attr': { priority: 'P1', wcagSC: '1.3.1', wcagLevel: 'A' },
        'th-has-data-cells': { priority: 'P1', wcagSC: '1.3.1', wcagLevel: 'A' },
        'valid-lang': { priority: 'P2', wcagSC: '3.1.1', wcagLevel: 'A' },
        'video-caption': { priority: 'P1', wcagSC: '1.2.2', wcagLevel: 'A' },
    })
);

export const base = getA11yConfig(baseRulesInfo);
