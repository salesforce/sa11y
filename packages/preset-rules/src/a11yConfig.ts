/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

import { RunOptions } from 'axe-core';

/**
 * A11yConfig defines options to run accessibility checks using axe specifying list of rules to test
 */
export interface A11yConfig extends RunOptions {
    runOnly: {
        type: 'rule';
        values: string[];
    };
    resultTypes: ['violations'];
}

/**
 * Returns config to be used in axe.run() with given rules
 *
 * @param rules - List of rules to be used in the config
 * @returns A11yConfig with formatted rules
 */
export function getA11yConfig(rules: string[]): A11yConfig {
    return {
        runOnly: {
            type: 'rule',
            values: rules,
        },
        resultTypes: ['violations'],
    };
}
