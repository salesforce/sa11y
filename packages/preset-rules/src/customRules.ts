/*
 * Copyright (c) 2024, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
import * as fs from 'fs';
import path from 'path';
import axe from 'axe-core';
import { processFiles } from '@sa11y/common';

// Directory containing custom rules and checks
const customRulesPathDir = 'custom-rules';

export function registerCustomRules(): void {
    const directoryPath = path.join(__dirname, customRulesPathDir);
    const newChecks: axe.Check[] = [];
    const newRules: axe.Rule[] = [];

    try {
        // Path to the JSON file with changes to existing rules
        const changesInExistingRulesPath = path.join(directoryPath, 'changes.json');
        // Read and parse existing rule changes
        const existingRuleData = fs.readFileSync(changesInExistingRulesPath, 'utf-8');
        const { rules } = JSON.parse(existingRuleData) as { rules: axe.Rule[] };
        if (rules && Array.isArray(rules)) {
            newRules.push(...rules);
        }

        // Process checks and rules JSON files
        processFiles<axe.Check>(path.join(directoryPath, 'checks'), newChecks, '.json', JSON.parse);
        processFiles<axe.Rule>(path.join(directoryPath, 'rules'), newRules, '.json', JSON.parse);
    } catch (e) {
        const err = e as Error;
        console.error('Error in reading Custom Rules files: ', err.message);
    }

    // Configure axe with the new checks and rules
    const spec: axe.Spec = { rules: newRules, checks: newChecks };
    axe.configure(spec);
}
