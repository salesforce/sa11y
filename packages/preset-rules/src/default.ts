/*
 * Copyright (c) 2021, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
import { A11yConfig } from '@sa11y/common';
import { base } from './base';
import { recommended } from './recommended';
import { full } from './full';

/**
 * Get ruleset from environment variable `SA11Y_RULESET` if defined.
 * Defaults to `recommended` ruleset.
 */
export function getDefaultRuleset(): A11yConfig {
    const ruleSets = new Map(Object.entries({ base: base, recommended: recommended, full: full }));
    const envRuleset = process.env.SA11Y_RULESET;
    return envRuleset ? ruleSets.get(envRuleset) || recommended : recommended;
}

export const defaultRuleset = getDefaultRuleset();
