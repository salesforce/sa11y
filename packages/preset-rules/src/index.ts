/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

export { defaultPriority, priorities, wcagLevels, getA11yConfig, RuleInfo } from './rules';
export { defaultRuleset, getDefaultRuleset } from './config';
export { extended } from './extended';
export { base } from './base';
export { full, excludedRules } from './full';
export { WcagMetadata } from './wcag';
export { registerCustomRules } from './customRules';
