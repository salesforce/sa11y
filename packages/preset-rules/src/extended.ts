/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

import { getA11yConfig } from './a11yConfig';
import { recommended } from './recommended';

const rules = [
    ...recommended.runOnly.values,
    'accesskeys',
    'aria-allowed-role',
    'avoid-inline-spacing',
    'css-orientation-lock',
    'frame-tested',
    'identical-links-same-purpose',
    'label-content-name-mismatch',
    'landmark-banner-is-top-level',
    'landmark-complementary-is-top-level',
    'landmark-contentinfo-is-top-level',
    'landmark-main-is-top-level',
    'landmark-no-duplicate-banner',
    'landmark-no-duplicate-contentinfo',
    'landmark-no-duplicate-main',
    'landmark-one-main',
    'landmark-unique',
    'link-in-text-block',
    'meta-viewport-large',
    'meta-viewport',
    'no-autoplay-audio',
    'object-alt',
    'p-as-heading',
    'region',
];

export const extended = getA11yConfig(rules);
