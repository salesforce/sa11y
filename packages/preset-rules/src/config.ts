/*
 * Copyright (c) 2021, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
import { A11yConfig } from '@sa11y/common';
import { base } from './base';
import { extended } from './extended';
import { full } from './full';

export const ruleSets = new Map(Object.entries({ base: base, extended: extended, full: full }));

/**
 * Get ruleset from environment variable `SA11Y_RULESET` if defined.
 * Defaults to `base` ruleset.
 */
export function getDefaultRuleset(): A11yConfig {
    const envRuleset = process.env.SA11Y_RULESET;
    return envRuleset ? ruleSets.get(envRuleset) || base : base;
}

export const defaultRuleset = getDefaultRuleset();
