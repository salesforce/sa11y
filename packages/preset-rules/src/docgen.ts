/*
 * Copyright (c) 2021, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
import { getAxeRules, log } from '@sa11y/common';
import { baseRulesInfo } from './base';
import { extendedRulesInfo } from './extended';
import * as fs from 'fs';
import { getMarkdownTable } from 'markdown-table-ts';

/**
 * Generate markdown table of rules with the rules metadata.
 * Called from the rules tests.
 * @param rulesInfo - preset-rule to generate docs for. Defaults to
 * @param updateReadmePath - append given file with generated rules md doc
 */
export function getRulesDoc(rulesInfo = extendedRulesInfo, updateReadmePath = ''): string {
    const rulesDocRows: string[][] = [];
    // Markers used to denote if a rule belongs in base, extended ruleset
    const no = '✖️';
    const yes = '✅';
    const axeRules = getAxeRules();
    for (const [ruleID, ruleMetadata] of rulesInfo.entries()) {
        const axeRule = axeRules.get(ruleID);
        if (!axeRule) throw new Error(`Unable to find rule: ${ruleID}`);

        rulesDocRows.push([
            `[${ruleID}](${axeRule.helpUrl.split('?')[0]})`, // remove URL referrer for cleaner/shorter doc
            axeRule.description.replace(/</g, '&lt;'),
            ruleMetadata.wcagSC,
            ruleMetadata.wcagLevel,
            ruleMetadata.priority,
            baseRulesInfo.has(ruleID) ? yes : no,
            yes,
        ]);
    }

    const table = getMarkdownTable({
        table: {
            head: [
                'Rule ID',
                'Description',
                'WCAG SC',
                'WCAG Level',
                'Priority',
                'In Base ruleset',
                'In Extended ruleset',
            ],
            body: [...rulesDocRows],
        },
    });

    // Note: Updating Readme is currently a manual dev process
    /* istanbul ignore next */
    if (updateReadmePath) {
        log(`Updating ${updateReadmePath} ..`);
        fs.appendFileSync(updateReadmePath, table);
    }

    return table;
}
