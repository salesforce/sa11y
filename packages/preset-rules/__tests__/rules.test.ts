/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

import axe from 'axe-core';
import { extended } from '../src';
import { recommended } from '../src';

/**
 * TODO:
 *  - Add benchmark test for individual rules using https://github.com/salesforce/best
 * */

describe('@sa11y/rules sanity checks', () => {
    it('recommended is a subset of extended', () => {
        expect(extended.runOnly.values).toEqual(expect.arrayContaining(recommended.runOnly.values));
        // TODO (debug): Why is this failing?
        // expect(extended).toEqual(expect.objectContaining(recommended));
    });
});

describe('@sa11y/rules sanity checks with axe', () => {
    // Rules that have been excluded from running due to being deprecated by axe
    // or due to their experimental nature
    const excludedRules = [
        'aria-dpub-role-fallback',
        'checkboxgroup',
        'frame-title-unique',
        'hidden-content',
        'layout-table',
        'radiogroup',
        'skip-link',
        'table-duplicate-name',
        'table-fake-caption',
        'video-description',
    ];
    const axeRules = axe.getRules().map((ruleObj) => ruleObj.ruleId);

    it('should not contain excluded, deprecated rules', () => {
        expect(extended.runOnly.values).toEqual(expect.not.arrayContaining(excludedRules));
    });

    it('all rules are present in axe', () => {
        expect(axeRules).toEqual(expect.arrayContaining(extended.runOnly.values));
    });

    it('only rules not used from axe are excluded, deprecated rules', () => {
        const unusedRules = axeRules.filter((rule) => !extended.runOnly.values.includes(rule));
        expect(unusedRules.sort()).toEqual(excludedRules.sort());
    });
});
