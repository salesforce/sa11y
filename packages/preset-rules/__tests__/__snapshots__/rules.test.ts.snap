// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`preset-rules documentation should throw error for non existent rule 1`] = `"Unable to find rule: foo"`;
